core/src/main/java/org/pentaho/di/core/parameters/NamedParamsDefault.java

```diff
@@ -0,0 +1,235 @@
+/*! ******************************************************************************
+ *
+ * Pentaho Data Integration
+ *
+ * Copyright (C) 2002-2017 by Hitachi Vantara : http://www.pentaho.com
+ *
+ *******************************************************************************
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.core.parameters;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * This class is an implementation of NamedParams.
+ *
+ * @author Sven Boden
+ */
+public class NamedParamsDefault implements NamedParams {
+  /**
+   * Map to store named parameters in.
+   */
+  protected Map<String, OneNamedParam> params = new HashMap<>();
+
+  /**
+   * Target class for the parameter keys.
+   */
+  class OneNamedParam {
+    /**
+     * key of this parameter
+     */
+    public String key;
+
+    /**
+     * Description of the parameter
+     */
+    public String description;
+
+    /**
+     * Default value for this parameter
+     */
+    public String defaultValue;
+
+    /**
+     * Actual value of the parameter.
+     */
+    public String value;
+  }
+
+  /**
+   * Default constructor.
+   */
+  public NamedParamsDefault() {
+  }
+
+  @Override
+  public void addParameterDefinition( String key, String defValue, String description ) throws DuplicateParamException {
+
+    if ( params.get( key ) == null ) {
+      OneNamedParam oneParam = new OneNamedParam();
+
+      oneParam.key = key;
+      oneParam.defaultValue = defValue;
+      oneParam.description = description;
+      oneParam.value = "";
+
+      params.put( key, oneParam );
+    } else {
+      throw new DuplicateParamException( "Duplicate parameter '" + key + "' detected." );
+    }
+  }
+
+  @Override
+  public String getParameterDescription( String key ) throws UnknownParamException {
+    String description = null;
+
+    OneNamedParam theParam = params.get( key );
+    if ( theParam != null ) {
+      description = theParam.description;
+    }
+
+    return description;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 91:98, 103:110, 115:122.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.